hypothesis/lms
# Improve grade validation accessibility using HTML5 attributes

> Closes #5205

This PR tries to improve accessibility around submitting invalid grades, by making the next changes:

* Convert the grade input to `type="number"`, and add the attributes `min={0}`, `max={scoreMaximum}` and `required`.
* Remove custom validation function, as everything it was checking is now covered by the attributes mentioned above.
* Remove the custom validation error toast message, as validation is now checked natively. This is more accessible, and screen readers will announce when the input value is invalid.

The main benefits of these changes is the amount of code we can get rid of, and the accessibility improvements already mentioned.

However, there's also some considerations and drawbacks:

* ~The comment popover cannot be closed when submitting the form. Otherwise, if the value was invalid, the re-render will make native validation popovers to disappear immediately (still checking if it's possible to prevent the input re-render in this case, and if that would actually make a difference).~ We have worked around this and is no longer a problem.
* ~The toast message was a bit nicer UI than native popovers. We could also consider styling invalid inputs here as a middle-ground solution.~ We decided it's ok to not use the message for now. If we want to get it back, we can use `ToastMessages` from frontend-shared, which is more accessible out of the box.

TODO:

- [x] Style input number to not display arrows.
- [x] Update frontend-shared to support `Input[type="number"]` (https://github.com/hypothesis/frontend-shared/pull/1248)
- [x] Remove CSS transition from validation toast message, if no longer needed.

### Testing steps

1. Check out this branch
2. Enable "grading comments" here http://localhost:8001/admin/instance/106/
3. Go to an assignment from that instance. For example https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/3348/View
4. Load a student
5. Try to leave the grade empty and click submit. It should not validate and warn you.
6. Try to set a grade smaller than 0 and click submit. It should not validate and warn you.
7. Try to set a value greater than the max score and click submit. It should not validate and warn you.
8. Repeat steps 5-7, but with the comment popover open, and using the submit button in there. Result should be the same, and the popover should not be closed.
9. Set a valid grade and click submit. It should successfully save it.
10. Repeat step 9, but with the comment popover open, and using the submit button in there. It should successfully save it and close the popover.

> This solution was first mentioned in https://github.com/hypothesis/lms/issues/5205#issuecomment-1731078936
